Boundary condition map

gassmoeller opened the pull request:
See changelog entry. This change allows the combination of plugins for boundary temperatures that is already possible for initial temperatures (I wish I had had this functionality for several of my papers). After review, I will apply the update_prm script on all included .prm files. For now I keep it out, because it makes the same one-line change in 500 files, which would make the review very tedious.

I plan to do a similar change for boundary_velocity and boundary_composition as well, but I would like to make sure I get the format right, before having to make the same edits to all of them. Boundary_fluid_pressure and Boundary_traction do not really have enough plugins as that a combination would make sense, but in some future we can think of that as well. Luckily the change is backwards compatible, as long as we keep the old single plugin input parameter around.

bangerth commented:
We seem to be accumulating these `Manager` classes. I wonder if we could reduce the amount of boiler-plate code if we had a class `Plugins::ManagerBase` or similar, templatized on the `Interface` class of a particular plugin system. 

That doesn't have to happen as part of this PR (and probably shouldn't), but I'd be curious about your opinion. I'm happy to help with this, that seems to be the kind of maintenance programming I have time for these days...

bangerth commented:
Yes, I also thought about using generic names. I'll get to it once the current patch is merged.

gassmoeller commented:
Addressed the comments. 
The sed command `/pattern1/,/pattern2/ { }` executes the rules in the braces in the range of lines between pattern1 and pattern2. In this case I want to make sure I am in the right subsection before replacing `Model name` because it appears in many subsections. Unfortunately I have not found a way to check for a matching `end` to the subsection command, so I limit the replacement to the first `end` encountered after the subsection. This might leave some parameters untouched (as described in the comment), but better to do nothing than something wrong.

bangerth commented:
Interesting command!

OK to merge once the tester is through.
